File: python_leetcode_2020/Python_Leetcode_2020/148_sort_list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    /**
     * Merge sort use bottom-up policy, 
     * so Space Complexity is O(1)
     * Time Complexity is O(NlgN)
     * stable sort
    */
    public ListNode sortList(ListNode head) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        int n = 0;
        while (head != null) {
            head = head.next;
            n++;
        }
        for (int step = 1; step < n; step <<= 1) {
            ListNode prev = dummy;
            ListNode cur = dummy.next;
            while (cur != null) {
                ListNode left = cur;
                ListNode right = split(left, step);
                cur = split(right, step);
                prev = merge(left, right, prev);
            } 
        }
        
        return dummy.next;
    }
    
    	/**
	 * Divide the linked list into two lists,
     * while the first list contains first n ndoes
	 * return the second list's head
	 */
    private ListNode split(ListNode head, int step) {
        if (head == null) return null;
    	
        for (int i = 1; head.next != null && i < step; i++) {
            head = head.next;
        }
        
        ListNode right = head.next;
        head.next = null;
        return right;
    }
    
    	/**
	  * merge the two sorted linked list l1 and l2,
	  * then append the merged sorted linked list to the node head
	  * return the tail of the merged sorted linked list
	 */
    private ListNode merge(ListNode left, ListNode right, ListNode prev) {
        ListNode cur = prev;
        while (left != null && right != null) {
            if (left.val < right.val) {
                cur.next = left;
                left = left.next;
            }
            else {
                cur.next = right;
                right = right.next;
            }
            cur = cur.next;
        }
        
        if (left != null) cur.next = left;
        else if (right != null) cur.next = right;
        while (cur.next != null) cur = cur.next;
        return cur;
    }
}

File: python_leetcode_2020/Python_Leetcode_2020/148_sort_list.py
"""
Sort a linked list in O(n log n) time using constant space complexity.
"""

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    # recursion
    def sortList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        #[4,2]
        if (head == None or head.next == None):
            return head

        # step 1. cut the list to two halves
        slow = head
        fast = head
        prev = None
        # second child list must start with slow pointer; 
        # this is because if there is only two elements in the list, slow point to the second element; 
        # to reduce the child list size, we have to split it before the slower pointer, otherwise we will get stuck with the same size and infinite loop
        while (fast != None and fast.next != None):
            prev = slow
            slow = slow.next
            fast = fast.next.next

        prev.next = None

        # step 2. sort each half
        l1 = self.sortList(head)
        l2 = self.sortList(slow)

        def merge(l1, l2):
            l = ListNode(0)
            p = l

            while (l1 != None and l2 != None):
                if (l1.val < l2.val):
                    p.next = l1
                    l1 = l1.next
                else:
                    p.next = l2
                    l2 = l2.next
                p = p.next

            if (l1 != None):
                p.next = l1

            if (l2 != None):
                p.next = l2

            return l.next
        
        # step 3. merge l1 and l2
        return merge(l1, l2)
